### PR TITLE
Create helper function to go through question answer pairs

### LINKED ISSUE #2618: Helper function to match the correct pair of qItem and qrItem
```
val qrItem = questionnaireResponse.allItems.find { qrItem -> qrItem.linkId == qItem.linkId }
```

this can get quite complicated in the world of nested items and nested repeated groups because link ids no longer uniquely identify a questionnaire item or a questionnaire response item...

i've been meaning to write a helper function that traverses the two trees in parallel and match the pairs.. maybe it's time for me to do that..

_Originally posted by @jingtang10 in https://github.com/google/android-fhir/pull/2522#discussion_r1671170333_

### PR DESCRIPTION
**Pull request by jingtang10**
**IMPORTANT: All PRs must be linked to an issue (except for extremely trivial and straightforward changes).**

Fixes #2618

**Description**
Clear and concise code change description.

**Alternative(s) considered**
Have you considered any alternatives? And if so, why have you chosen the approach in this PR?

**Type**
Choose one: (Bug fix | Feature | Documentation | Testing | Code health | Builds | Releases | Other)

**Screenshots (if applicable)**

**Checklist**

- [x] I have read and acknowledged the [Code of conduct](https://github.com/google/android-fhir/blob/master/CODE_OF_CONDUCT.md).
- [x] I have read the [Contributing](https://google.github.io/android-fhir/contrib/) page.
- [x] I have signed the Google [Individual CLA](https://cla.developers.google.com/about/google-individual), or I am covered by my company's [Corporate CLA](https://cla.developers.google.com/about/google-corporate ).
- [x] I have discussed my proposed solution with code owners in the linked issue(s) and we have agreed upon the general approach.
- [x] I have run `./gradlew spotlessApply` and `./gradlew spotlessCheck` to check my code follows the style guide of this project.
- [x] I have run `./gradlew check` and `./gradlew connectedCheck` to test my changes locally.
- [x] I have built and run the demo app(s) to verify my change fixes the issue and/or does not break the demo app(s).

### CODE REVIEW

**Comment by jingtang10** [comment]
@FikriMilano please take a look at this

File: datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaires.kt [hunk 1]
```diff
@@ -163,3 +166,77 @@ enum class EntryMode(val value: String) {
...
+/**
+ * Applies `forEach` on each questionnaire item and questionnaire response item pair in the
+ * questionnaire and the given `questionnaireResponse`.
+ *
+ * Questionnaire items and questionnaire response items are visited in pre-order.
+ *
+ * Items nested under repeated groups and repeated questions will be repeated for each repeated
+ * group instance or answer provided by the user.
+ */
+internal suspend fun Questionnaire.forEachItemPair(
```
[the remark this answers is not in the document]
Comment (jingtang10) [reply datacapture/src/main/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnaires.kt:182]: done

**Review by jingtang10 — COMMENTED** [review]
Thanks @FikriMilano

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt [hunk 2]
```diff
@@ -240,4 +245,270 @@ class MoreQuestionnairesTest {
         "The extension:name is missing or is not of type Coding in $EXTENSION_SDC_QUESTIONNAIRE_LAUNCH_CONTEXT",
       )
   }
+
+  @Test
+  fun `forEachPair should throw an exception if questionnaire and questionnaire response item numbers do not match`() =
```
Comment (MJ1998) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt:250]:
```suggestion
  fun `forEachItemPair should throw an exception if questionnaire and questionnaire response item numbers do not match`() =
```

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt [hunk 3]
```diff
@@ -240,4 +245,270 @@ class MoreQuestionnairesTest {
...
+          )
+        }
+
+      assertFailsWith<IllegalArgumentException> {
+        questionnaire.forEachItemPair(questionnaireResponse) { _, _ -> }
+      }
+    }
+
+  @Test
+  fun `forEachPair should throw an exception if questionnaire and questionnaire response link ids do not match`() =
```
Comment (MJ1998) [inline datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt:290]: Similarly here and other places

**Review by MJ1998 — APPROVED** [review]
Nits. LGTM

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt (hunk 2 above)
Comment (jingtang10) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt:250]: done

File: datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt (hunk 3 above)
Comment (jingtang10) [reply datacapture/src/test/java/com/google/android/fhir/datacapture/extensions/MoreQuestionnairesTest.kt:290]: done

**Pull request merged**